workflow/lib/sbs_process/eval.py

```diff
@@ -0,0 +1,584 @@
+            y="mapping_rate",
+            **kwargs,
+        )
+    else:
+        sns.lineplot(
+            data=df_summary,
+            x="{}_threshold".format(threshold_var),
+            y="mapping_rate",
+            ax=ax,
+            **kwargs,
+        )
+    ax.set_ylabel("mapping rate", fontsize=18)
+    ax.set_xlabel("{} threshold".format(threshold_var), fontsize=18)
+    ax_right = ax.twinx()
+    sns.lineplot(
+        data=df_summary,
+        x="{}_threshold".format(threshold_var),
+        y="mapped_spots",
+        ax=ax_right,
+        color="coral",
+        **kwargs,
+    )
+    ax_right.set_ylabel("mapped spots", fontsize=18)
+    plt.legend(
+        ax.get_lines() + ax_right.get_lines(), ["mapping rate", "mapped spots"], loc=7
+    )
+
+    return df_summary
+
+
+def plot_read_mapping_heatmap(
+    df_reads,
+    barcodes,
+    shape="square",
+    plate="6W",
+    return_plot=True,
+    return_summary=False,
+    **kwargs,
+):
+    """Plot the mapping rate of reads by well and tile in a convenient plate layout.
+
+    Args:
+        df_reads (pandas.DataFrame):
+            DataFrame of all reads output from sbs mapping pipeline, e.g., concatenated outputs for all tiles
+            and wells of Snake.call_reads().
+        barcodes (list or set of str):
+            Expected barcodes from the pool library design.
+        shape (str, optional):
+            Shape of subplot for each well used in plot_plate_heatmap. Defaults to 'square'.
+        plate (str):
+            Plate type for plot_plate_heatmap. Options are {'6W', '24W', '96W'}.
+        return_plot (bool, optional):
+            If true, returns df_summary. Defaults to True.
+        return_summary (bool, optional):
+            If true, returns df_summary. Defaults to False.
+        **kwargs:
+            Keyword arguments passed to plot_plate_heatmap().
+
+    Returns:
+        pandas.DataFrame: DataFrame used for plotting, optional output, only returns if return_summary=True.
+        np.array: Array of matplotlib Axes objects.
+    """
+    # Mark reads as mapped or unmapped based on provided barcodes
+    df_reads.loc[:, "mapped"] = df_reads["barcode"].isin(barcodes)
+
+    # Calculate mapping rates by well and tile
+    df_summary = (
+        df_reads.groupby(["well", "tile"])["mapped"]
+        .value_counts(normalize=True)
+        .rename("fraction of reads mapping")
+        .to_frame()
+        .reset_index()
+        .query("mapped")
+        .drop(columns="mapped")
+    )
+
+    if return_summary and return_plot:
+        # Plot heatmap
+        axes = plot_plate_heatmap(df_summary, shape=shape, plate=plate, **kwargs)
+        return df_summary, axes
+    elif return_plot:
+        # Plot heatmap
+        axes = plot_plate_heatmap(df_summary, shape=shape, plate=plate, **kwargs)
+        return axes
+    elif return_summary:
+        return df_summary
+    else:
+        return None
+
+
+def plot_plate_heatmap(
```

Comment:
This is a shared function. It is used in other steps as well!

Reply:
Thanks for pointing this out! Done in https://github.com/cheeseman-lab/brieflow/pull/10/commits/f4d9c741ca2cdb6a4951b6a1c3e4a24e6035e89e